river/Touch.zig

```diff
@@ -0,0 +1,109 @@
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+const Self = @This();
+
+const std = @import("std");
+const wlr = @import("wlroots");
+const wayland = @import("wayland");
+const wl = wayland.server.wl;
+const zwlr = wayland.server.zwlr;
+
+const c = @import("c.zig");
+const server = &@import("main.zig").server;
+const util = @import("util.zig");
+
+const Cursor = @import("Cursor.zig");
+const Seat = @import("Seat.zig");
+
+const log = std.log.scoped(.touch);
+
+seat: *Seat,
+wlr_cursor: *wlr.Cursor,
+
+touch_up: wl.Listener(*wlr.Touch.event.Up) =
+    wl.Listener(*wlr.Touch.event.Up).init(handleTouchUp),
+touch_down: wl.Listener(*wlr.Touch.event.Down) =
+    wl.Listener(*wlr.Touch.event.Down).init(handleTouchDown),
+touch_motion: wl.Listener(*wlr.Touch.event.Motion) =
+    wl.Listener(*wlr.Touch.event.Motion).init(handleTouchMotion),
+
+pub fn init(self: *Self, seat: *Seat) !void {
+    const wlr_cursor = try wlr.Cursor.create();
+    errdefer wlr_cursor.destroy();
+    wlr_cursor.attachOutputLayout(server.root.output_layout);
+
+    self.* = .{
+        .seat = seat,
+        .wlr_cursor = wlr_cursor,
+    };
+
+    wlr_cursor.events.touch_up.add(&self.touch_up);
+    wlr_cursor.events.touch_down.add(&self.touch_down);
+    wlr_cursor.events.touch_motion.add(&self.touch_motion);
+}
+
+pub fn deinit(self: *Self) void {
+    self.wlr_cursor.destroy();
+}
+
+fn handleTouchMotion(
+    listener: *wl.Listener(*wlr.Touch.event.Motion),
+    event: *wlr.Touch.event.Motion,
+) void {
+    const self = @fieldParentPtr(Self, "touch_motion", listener);
+
+    self.seat.handleActivity();
+
+    var lx: f64 = undefined;
+    var ly: f64 = undefined;
+    self.wlr_cursor.absoluteToLayoutCoords(event.device, event.x, event.y, &lx, &ly);
+
+    if (Cursor.surfaceAtPosition(lx, ly)) |result| {
+        _ = self.seat.wlr_seat.touchNotifyMotion(event.time_msec, event.touch_id, result.sx, result.sy);
+    }
+}
+
+fn handleTouchUp(
+    listener: *wl.Listener(*wlr.Touch.event.Up),
+    event: *wlr.Touch.event.Up,
+) void {
+    const self = @fieldParentPtr(Self, "touch_up", listener);
+
+    self.seat.handleActivity();
+
+    _ = self.seat.wlr_seat.touchNotifyUp(event.time_msec, event.touch_id);
+}
+
+fn handleTouchDown(
+    listener: *wl.Listener(*wlr.Touch.event.Down),
+    event: *wlr.Touch.event.Down,
+) void {
+    const self = @fieldParentPtr(Self, "touch_down", listener);
+
+    self.seat.handleActivity();
+
+    var lx: f64 = undefined;
+    var ly: f64 = undefined;
+    self.wlr_cursor.absoluteToLayoutCoords(event.device, event.x, event.y, &lx, &ly);
+
+    if (Cursor.surfaceAtPosition(lx, ly)) |result| {
+        _ = self.seat.wlr_seat.touchNotifyDown(result.surface, event.time_msec, event.touch_id, result.sx, result.sy);
+    }
+
+    self.seat.cursor.hide();
```

Comment:
Why hide the main mouse cursor? It has nothing to do with the touch operation.

Reply:
It's something that annoys me about GNOME. When switching my laptop to tablet mode, it's a bit annoying to have a cursor in the middle of my screen that I can't move in any way because I can't use the mouse. Sway and KDE do hide the mouse cursor when using touch, but I understand that it seems a bit odd at first.